### PR TITLE
Add loading indicator to the test liveview

### PR DESCRIPTION
**Pull request by r-richardson**
This PR adds a loading animation indicating that the liveview is being loaded. Currently only a empty box is displayed, visible in the webui before liveview content is received by the responsible canvas element.
Test cases are also added to confirm the elements are displayed and hidden as expected.

Related Issue: https://progress.opensuse.org/issues/134840

loading_animation

### CODE REVIEW

File: assets/javascripts/running.js [hunk 1]
```diff
@@ -321,18 +321,29 @@ var last_event;
...
   };
   scrn.src = dataURL;
 }
 
 function initLivestream() {
+  // Show the loading animation when initializing the live stream
+  var loadingElement = document.getElementById('liveview-loading');
+  if (loadingElement) {
+    loadingElement.style.display = 'block';
+  }
```
Comment (okurz) [inline assets/javascripts/running.js:346]: Unless you show me in a test I don't believe this should be necessary at all. Proof please :)

**Review by okurz — CHANGES_REQUESTED** [review]

**Comment by kalikiana** [comment]
```
[15:22:13] t/full-stack.t ......... 31/? Bailout called.  Further testing stopped:  test 't/full-stack.t' exceeds runtime limit of '1800' seconds
FAILED--Further testing stopped: test 't/full-stack.t' exceeds runtime limit of '1800' seconds
```

File: assets/javascripts/running.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (r-richardson) [reply assets/javascripts/running.js:346]: yes you're right, this part is probably not be needed. I'm trying to verify that with tests atm

File: assets/stylesheets/result_preview.scss [hunk 2]
```diff
@@ -91,7 +91,6 @@
     // live view
     canvas {
         max-width: 100%;
-        border: #ddd 1px solid;
```
Comment (okurz) [inline assets/stylesheets/result_preview.scss:94]: I think the border can give some good indication what's going on if there's a problem so consider to keep that

File: t/ui/18-tests-details.t [hunk 3]
```diff
@@ -291,6 +291,7 @@ sub update_status {
 }
 
 subtest 'running job' => sub {
+
```
Comment (okurz) [inline t/ui/18-tests-details.t:294]: Why this unrelated change?

File: t/ui/18-tests-details.t [hunk 4]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
     # waiting for the periodic call anyways
     $driver->execute_script('window.enableStatusUpdates = false');
 
+    subtest 'liveview loading animation is displayed' => sub {
+        # ensure we are on the Liveview tab
+        like(current_tab, qr/live/i, 'live tab active by default');
+
+        # switch to another tab (e.g., Details)
```
Comment (okurz) [inline t/ui/18-tests-details.t:319]: Redundant comment. The test string is already stating that

File: t/ui/18-tests-details.t [hunk 5]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
     # waiting for the periodic call anyways
     $driver->execute_script('window.enableStatusUpdates = false');
 
+    subtest 'liveview loading animation is displayed' => sub {
+        # ensure we are on the Liveview tab
```
Comment (okurz) [inline t/ui/18-tests-details.t:316]: Redundant comment. The test string *could* state that

File: t/ui/18-tests-details.t [hunk 6]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        like(current_tab, qr/live/i, 'live tab active by default');
+
+        # switch to another tab (e.g., Details)
+        $driver->find_element_by_link_text('Details')->click();
+        like(current_tab, qr/details/i, 'switched to details tab');
+
+        # switch back to Live View tab
```
Comment (okurz) [inline t/ui/18-tests-details.t:323]: Redundant

File: t/ui/18-tests-details.t [hunk 7]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        $driver->find_element_by_link_text('Live View')->click();
+        like(current_tab, qr/live/i, 'switched back to live tab');
+
+        # check that the liveview-loading div exists after tab switch
```
Comment (okurz) [inline t/ui/18-tests-details.t:327]: Redundant

File: t/ui/18-tests-details.t [hunk 8]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        my $loading_element = $driver->find_element('#liveview-loading', 'css');
+        ok($loading_element, 'liveview-loading element exists after tab switch');
+
+        # check that the liveview-loading div is visible after tab switch
```
Comment (okurz) [inline t/ui/18-tests-details.t:331]: Redundant

File: t/ui/18-tests-details.t [hunk 9]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        ok($loading_element->is_displayed(), 'liveview-loading is visible after tab switch');
+    };
+
+    subtest 'liveview loading animation hides after live view starts' => sub {
+        # ensure the liveview-loading element is initially visible
```
Comment (okurz) [inline t/ui/18-tests-details.t:336]: Redundant

File: t/ui/18-tests-details.t [hunk 10]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        my $loading_element = $driver->find_element('#liveview-loading');
+        ok($loading_element->is_displayed(), 'liveview-loading is initially visible');
+
+        # simulate the live view starting
```
Comment (okurz) [inline t/ui/18-tests-details.t:340]: *that* is a good comment unlike the examples before as is provides new information and explains the not obvious next part, keep this

File: t/ui/18-tests-details.t [hunk 11]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        $driver->execute_script(
+            q{
+            var canvas = document.getElementById('livestream');
+            var dataURL = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR4nGNgYAAAAAMAAWgmWQ0AAAAASUVORK5CYII=';
+            loadCanvas(canvas, dataURL);
+        }
+        );
+        # wait until the loading animation is hidden
```
Comment (okurz) [inline t/ui/18-tests-details.t:348]: Redundant

File: t/ui/18-tests-details.t [hunk 12]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        wait_until sub {
+            my $loading_element = $driver->find_element('#liveview-loading');
+            !$loading_element->is_displayed();
+        }, 'liveview-loading is hidden after live view starts', 5;
+        # verify that the canvas is now displayed
+        my $canvas = $driver->find_element('#livestream');
+        ok($canvas->is_displayed(), 'livestream canvas is displayed');
+
+        # switch to another tab (e.g., Details)
```
Comment (okurz) [inline t/ui/18-tests-details.t:357]: .

File: t/ui/18-tests-details.t [hunk 13]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        $driver->find_element_by_link_text('Details')->click();
+        like(current_tab, qr/details/i, 'switched to details tab');
+
+        # switch back to Live View tab
```
Comment (okurz) [inline t/ui/18-tests-details.t:361]: .

File: t/ui/18-tests-details.t [hunk 14]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        $driver->find_element_by_link_text('Live View')->click();
+        like(current_tab, qr/live/i, 'switched back to live tab');
+
+        # check that the liveview-loading div exists after tab switch
```
Comment (okurz) [inline t/ui/18-tests-details.t:365]: .

File: t/ui/18-tests-details.t [hunk 15]
```diff
@@ -311,6 +312,63 @@ subtest 'running job' => sub {
...
+        $loading_element = $driver->find_element('#liveview-loading', 'css');
+        ok($loading_element, 'liveview-loading element exists after tab switch');
+
+        # check that the liveview-loading div is still hidden after tab switch
```
Comment (okurz) [inline t/ui/18-tests-details.t:369]: .

**Review by okurz — CHANGES_REQUESTED** [review]

File: assets/stylesheets/result_preview.scss (hunk 2 above)
Comment (r-richardson) [reply assets/stylesheets/result_preview.scss]: yes, though it is then visible during loading
canvas_border
i tried including the loading animation within the canvas, though that hid it.

**Review by okurz — DISMISSED** [review]

File: t/ui/18-tests-details.t
Comment (okurz) [inline t/ui/18-tests-details.t:1]: I tried out your branch and ran `OPENQA_TEST_TIMEOUT_DISABLE=1 count-fail-ratio prove -l t/ui/18-tests-details.t` and found a sporadic failure, reproduced only once so far:

```
## Run 1
t/ui/18-tests-details.t .. ok    
All tests successful.
Files=1, Tests=36, 43 wallclock secs ( 0.09 usr  0.02 sys + 10.22 cusr  0.89 csys = 11.22 CPU)
Result: PASS
## count-fail-ratio: Run: 1. Fails: 0. Fail ratio 0±0%. No fails, computed failure probability < 300.00%
## mean runtime: 42950±0 ms
…
## Run 12
t/ui/18-tests-details.t .. 19/? isElementEnabled: stale element reference: stale element not found in the current frame at /home/okurz/local/os-autoinst/openQA/t/ui/../lib/OpenQA/SeleniumTest.pm:78 at /home/okurz/local/os-autoinst/openQA/t/ui/../lib/OpenQA/SeleniumTest.pm line 81.
	OpenQA::SeleniumTest::__ANON__(Test::Selenium::Chrome=HASH(0x55feb6a44540), "Error while executing command: isElementEnabled: stale elemen"..., HASH(0x55feb7dc5d80)) called at /usr/lib/perl5/vendor_perl/5.26.1/Selenium/Remote/Driver.pm line 356
	Selenium::Remote::Driver::catch {...} ("Error while executing command: isElementEnabled: stale elemen"...) called at /usr/lib/perl5/vendor_perl/5.26.1/Try/Tiny.pm line 123
	Try::Tiny::try(CODE(0x55feb7bb82c0), Try::Tiny::Catch=REF(0x55feb7e41c58)) called at /usr/lib/perl5/vendor_perl/5.26.1/Selenium/Remote/Driver.pm line 361
	Selenium::Remote::Driver::__ANON__(CODE(0x55feb65f7508), Test::Selenium::Chrome=HASH(0x55feb6a44540), HASH(0x55feb7dc5d80)) called at (eval 1582) line 1
	Selenium::Remote::Driver::__ANON__(Test::Selenium::Chrome=HASH(0x55feb6a44540), HASH(0x55feb7dc5d80)) called at (eval 1584) line 2
	Selenium::Remote::Driver::_execute_command(Test::Selenium::Chrome=HASH(0x55feb6a44540), HASH(0x55feb7dc5d80)) called at (eval 1554) line 17
	Selenium::Remote::WebElement::_execute_command(Test::Selenium::Remote::WebElement=HASH(0x55feb79fb418), HASH(0x55feb7dc5d80)) called at /usr/lib/perl5/vendor_perl/5.26.1/Selenium/Remote/WebElement.pm line 180
	Selenium::Remote::WebElement::is_enabled(Test::Selenium::Remote::WebElement=HASH(0x55feb79fb418)) called at t/ui/18-tests-details.t line 84
	main::find_candidate_needles() called at t/ui/18-tests-details.t line 654
	main::test_with_error(1, 0, ARRAY(0x55feb7e46a90), HASH(0x55feb7e3bd08), "100%, 52%") called at t/ui/18-tests-details.t line 674
	main::__ANON__() called at /usr/lib/perl5/vendor_perl/5.26.1/Test/Builder.pm line 374
	eval {...} called at /usr/lib/perl5/vendor_perl/5.26.1/Test/Builder.pm line 374
	Test::Builder::subtest(Test::Builder=HASH(0x55feac77c2a0), "test candidate list", CODE(0x55feb768dda8)) called at /usr/lib/perl5/vendor_perl/5.26.1/Test/More.pm line 831
	Test::More::subtest("test candidate list", CODE(0x55feb768dda8)) called at t/ui/18-tests-details.t line 706
t/ui/18-tests-details.t .. 20/? # Tests were run but no plan was declared and done_testing() was not seen.
t/ui/18-tests-details.t .. Dubious, test returned 254 (wstat 65024, 0xfe00)
All 20 subtests passed 
Test Summary Report
-------------------
t/ui/18-tests-details.t (Wstat: 65024 Tests: 20 Failed: 0)
  Non-zero exit status: 254
  Parse errors: No plan found in TAP output
Files=1, Tests=20, 34 wallclock secs ( 0.07 usr  0.00 sys +  8.60 cusr  1.06 csys =  9.73 CPU)
Result: FAIL
## count-fail-ratio: Run: 12. Fails: 1. Fail ratio 8.33±15.63%
## mean runtime: 39634±2031.91 ms
```

Comment (r-richardson) [reply t/ui/18-tests-details.t:1]: yes, this is the same error i've also been running into. Is this maybe related to the line
[tests-details.t:311](https://github.com/os-autoinst/openQA/pull/6023/files#diff-752f62785dc57e09d41a9dd89bdc34e67f66a81de9248dda0482be71e8ab4d3aL312) `$driver->execute_script('window.enableStatusUpdates = false');` ?
I tried to move this line directly after my tests, though that did not help.

Comment (okurz) [reply t/ui/18-tests-details.t:1]: By the way I compared that to origin/master. Result:

```
## count-fail-ratio: Run: 30. Fails: 0. Fail ratio 0±0%. No fails, computed failure probability < 10.00%
## mean runtime: 36969±1693.07 ms
```

so it's very likely error you see is caused by your change.

File: assets/stylesheets/result_preview.scss (hunk 2 above)
Comment (Martchus) [reply assets/stylesheets/result_preview.scss]: > i tried including the loading animation within the canvas, though that hid it.

Maybe it helps to assign a high z-index to the loading animation.

File: assets/javascripts/running.js [hunk 16]
```diff
@@ -321,13 +321,19 @@ var last_event;
...
+  // load image from data URL
   var scrn = new Image();
   scrn.onload = function () {
     canvas.width = this.width;
     canvas.height = this.height;
     context.clearRect(0, 0, this.width, this.width);
     context.drawImage(this, 0, 0);
+
+    // hide loading animation after the first image is loaded
+    var loadingElement = document.getElementById('liveview-loading');
```
Comment (Martchus) [inline assets/javascripts/running.js:333]:
```suggestion
    const loadingElement = document.getElementById('liveview-loading');
```

File: assets/javascripts/running.js [hunk 17]
```diff
@@ -321,13 +321,19 @@ var last_event;
...
+    if (loadingElement) {
+      loadingElement.style.display = 'none';
+    }
```
Comment (Martchus) [inline assets/javascripts/running.js:336]: The element is likely always present at the point this callback is invoked. You could check whether that's the case and if it is the case remove the condition.

File: t/ui/18-tests-details.t
Comment (r-richardson) [reply t/ui/18-tests-details.t:1]: Turned out we both didn't try often enough, as master was affected after all.
There where some missing wait_for_ajax calls, @perlpunk was able to quickly find and [fix the issue.](https://github.com/os-autoinst/openQA/pull/6040)

**Review by asdil12 — APPROVED** [review]
